File: bindings/prolog/safe_network.pl
%% SPDX-License-Identifier: PMPL-1.0-or-later
%% Copyright (c) 2026 Jonathan D.A. Jewell (hyperpolymath) <jonathan.jewell@open.ac.uk>
%%
%% Proven SafeNetwork - Extended FFI bindings to libproven network operations.
%% All network operations are performed in verified Idris 2 code via libproven.

:- module(safe_network, [
    parse_ipv4/2,
    is_loopback/1,
    is_private_ip/1,
    is_public_ip/1,
    classify_ip/2,
    is_valid_port/1,
    is_privileged_port/1
]).

:- use_module(proven_network, [
    parse_ipv4/2 as pn_parse_ipv4,
    is_loopback/1 as pn_is_loopback,
    is_private_ip/1 as pn_is_private_ip,
    is_public_ip/1 as pn_is_public_ip,
    classify_ip/2 as pn_classify_ip,
    is_valid_port/1 as pn_is_valid_port,
    is_privileged_port/1 as pn_is_privileged_port
]).

%! parse_ipv4(+IpString, -Ip) is det.
%
%  Parse IPv4 address string via libproven.
parse_ipv4(IpString, Ip) :-
    pn_parse_ipv4(IpString, Ip).

%! is_loopback(+Ip) is semidet.
%
%  Check if IPv4 address is loopback via libproven.
is_loopback(Ip) :-
    pn_is_loopback(Ip).

%! is_private_ip(+Ip) is semidet.
%
%  Check if IPv4 address is private via libproven.
is_private_ip(Ip) :-
    pn_is_private_ip(Ip).

%! is_public_ip(+Ip) is semidet.
%
%  Check if IPv4 address is public via libproven.
is_public_ip(Ip) :-
    pn_is_public_ip(Ip).

%! classify_ip(+Ip, -Class) is det.
%
%  Classify an IPv4 address via libproven.
classify_ip(Ip, Class) :-
    pn_classify_ip(Ip, Class).

%! is_valid_port(+Port) is semidet.
%
%  Check if port number is valid (1-65535).
%  Uses proven_network for validation.
is_valid_port(Port) :-
    pn_is_valid_port(Port).

%! is_privileged_port(+Port) is semidet.
%
%  Check if port is a privileged port (1-1023).
is_privileged_port(Port) :-
    pn_is_privileged_port(Port).

File: bindings/purescript/Proven/SafeNetwork.js
// SPDX-License-Identifier: PMPL-1.0-or-later
// Copyright (c) 2026 Jonathan D.A. Jewell (hyperpolymath) <jonathan.jewell@open.ac.uk>
//
// SafeNetwork FFI implementation - delegates to libproven via ffi-napi.
// All computation performed in Idris 2 via the Zig FFI layer.

"use strict";

const { lib } = require("./FFI.js");

exports.parseIpv4Impl = function (s) {
  var buf = Buffer.from(s, "utf-8");
  var r = lib.proven_network_parse_ipv4(buf, buf.length);
  if (r.status === 0) {
    return {
      status: 0,
      octets: [
        r.address.octets[0],
        r.address.octets[1],
        r.address.octets[2],
        r.address.octets[3],
      ],
    };
  }
  return { status: r.status, octets: [0, 0, 0, 0] };
};

exports.ipv4IsPrivateImpl = function (octets) {
  var addr = { octets: octets };
  return lib.proven_network_ipv4_is_private(addr);
};

exports.ipv4IsLoopbackImpl = function (octets) {
  var addr = { octets: octets };
  return lib.proven_network_ipv4_is_loopback(addr);
};
